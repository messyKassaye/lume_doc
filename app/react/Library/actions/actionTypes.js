export const SET_DOCUMENTS = 'SET_DOCUMENTS';
export const ADD_DOCUMENTS = 'ADD_DOCUMENTS';
export const UNSET_DOCUMENTS = 'UNSET_DOCUMENTS';
export const UPDATE_DOCUMENT = 'LIBRARY/UPDATE_DOCUMENT';
export const UPDATE_DOCUMENTS = 'LIBRARY/UPDATE_DOCUMENTS';
export const SET_LIBRARY_TEMPLATES = 'SET_LIBRARY_TEMPLATES';
export const ELEMENT_CREATED = 'ELEMENT_CREATED';
export const REMOVE_DOCUMENT = 'REMOVE_DOCUMENT';
export const REMOVE_DOCUMENTS = 'REMOVE_DOCUMENTS';
export const UPDATE_SELECTED_ENTITIES = 'UPDATE_SELECTED_ENTITIES';
//Search
export const SET_SEARCHTERM = 'SET_SEARCHTERM';
export const SET_PREVIEW_DOC = 'SET_PREVIEW_DOC';
export const SET_SUGGESTIONS = 'SET_SUGGESTIONS';
export const SHOW_SUGGESTIONS = 'SHOW_SUGGESTIONS';
export const HIDE_SUGGESTIONS = 'HIDE_SUGGESTIONS';
export const OVER_SUGGESTIONS = 'OVER_SUGGESTIONS';
//Filters
export const SET_LIBRARY_FILTERS = 'SET_LIBRARY_FILTERS';
export const UPDATE_LIBRARY_FILTERS = 'UPDATE_LIBRARY_FILTERS';
export const INITIALIZE_FILTERS_FORM = 'INITIALIZE_FILTERS_FORM';
//ui
export const SELECT_DOCUMENT = 'SELECT_DOCUMENT';
export const SELECT_SINGLE_DOCUMENT = 'SELECT_SINGLE_DOCUMENT';
export const SELECT_DOCUMENTS = 'SELECT_DOCUMENTS';
export const UNSELECT_DOCUMENT = 'UNSELECT_DOCUMENT';
export const UNSELECT_ALL_DOCUMENTS = 'UNSELECT_ALL_DOCUMENTS';
export const SHOW_FILTERS = 'SHOW_FILTERS';
export const HIDE_FILTERS = 'HIDE_FILTERS';
export const ENTER_LIBRARY = 'ENTER_LIBRARY';
export const ZOOM_IN = 'ZOOM_IN';
export const ZOOM_OUT = 'ZOOM_OUT';
export const SET_TABLE_VIEW_COLUMNS = 'SET_TABLE_VIEW_COLUMNS';
export const SET_TABLE_VIEW_COLUMN_HIDDEN = 'SET_TABLE_VIEW_COLUMN_HIDDEN';
export const SET_TABLE_VIEW_ALL_COLUMNS_HIDDEN = 'SET_TABLE_VIEW_ALL_COLUMNS_HIDDEN';
