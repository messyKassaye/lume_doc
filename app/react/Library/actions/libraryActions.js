import * as types from 'app/Library/actions/actionTypes';
import api from 'app/Search/SearchAPI';
import { notificationActions } from 'app/Notifications';
import { actions as formActions } from 'react-redux-form';
import { actions } from 'app/BasicReducer';
import { documentsApi } from 'app/Documents';
import { browserHistory } from 'react-router';
import rison from 'rison-node';
import referencesAPI from 'app/Viewer/referencesAPI';
import { api as entitiesAPI } from 'app/Entities';
import { toUrlParams } from 'shared/JSONRequest';
import { RequestParams } from 'app/utils/RequestParams';
import { store } from 'app/store';
import searchAPI from 'app/Search/SearchAPI';
import { selectedDocumentsChanged, maybeSaveQuickLabels } from './quickLabelActions';

export function enterLibrary() {
  return { type: types.ENTER_LIBRARY };
}

export function initializeFiltersForm(values = {}) {
  return Object.assign(values, { type: types.INITIALIZE_FILTERS_FORM });
}

export function selectDocument(_doc) {
  return async (dispatch, getState) => {
    const doc = _doc.toJS ? _doc.toJS() : _doc;
    const showingSemanticSearch = getState().library.sidepanel.tab === 'semantic-search-results';
    if (showingSemanticSearch && !doc.semanticSearch) {
      dispatch(actions.set('library.sidepanel.tab', ''));
    }
    await dispatch(maybeSaveQuickLabels());
    dispatch({ type: types.SELECT_DOCUMENT, doc });
    dispatch(selectedDocumentsChanged());
  };
}

export function getAndSelectDocument(sharedId) {
  return dispatch => {
    entitiesAPI.get(new RequestParams({ sharedId })).then(entity => {
      dispatch({ type: types.SELECT_SINGLE_DOCUMENT, doc: entity[0] });
    });
  };
}

export function selectDocuments(docs) {
  return async dispatch => {
    await dispatch(maybeSaveQuickLabels());
    dispatch({ type: types.SELECT_DOCUMENTS, docs });
    dispatch(selectedDocumentsChanged());
  };
}

export function unselectDocument(docId) {
  return async dispatch => {
    await dispatch(maybeSaveQuickLabels());
    dispatch({ type: types.UNSELECT_DOCUMENT, docId });
    dispatch(selectedDocumentsChanged());
  };
}

export function selectSingleDocument(doc) {
  return async dispatch => {
    await dispatch(maybeSaveQuickLabels());
    dispatch({ type: types.SELECT_SINGLE_DOCUMENT, doc });
    dispatch(selectedDocumentsChanged());
  };
}

export function unselectAllDocuments() {
  return async dispatch => {
    await dispatch(maybeSaveQuickLabels());
    dispatch({ type: types.UNSELECT_ALL_DOCUMENTS });
    dispatch(selectedDocumentsChanged());
  };
}

export function updateSelectedEntities(entities) {
  return { type: types.UPDATE_SELECTED_ENTITIES, entities };
}

export function showFilters() {
  return { type: types.SHOW_FILTERS };
}

export function hideFilters() {
  return { type: types.HIDE_FILTERS };
}

export function setDocuments(docs) {
  return { type: types.SET_DOCUMENTS, documents: docs };
}

export function addDocuments(docs) {
  return { type: types.ADD_DOCUMENTS, documents: docs };
}

export function unsetDocuments() {
  return { type: types.UNSET_DOCUMENTS };
}

export function setTemplates(templates, thesauris) {
  return dispatch => {
    dispatch({ type: types.SET_LIBRARY_TEMPLATES, templates, thesauris });
  };
}

export function setPreviewDoc(docId) {
  return { type: types.SET_PREVIEW_DOC, docId };
}

export function setSuggestions(suggestions) {
  return { type: types.SET_SUGGESTIONS, suggestions };
}

export function hideSuggestions() {
  return { type: types.HIDE_SUGGESTIONS };
}

export function showSuggestions() {
  return { type: types.SHOW_SUGGESTIONS };
}

export function setOverSuggestions(boolean) {
  return { type: types.OVER_SUGGESTIONS, hover: boolean };
}

export function zoomIn() {
  return { type: types.ZOOM_IN };
}

export function zoomOut() {
  return { type: types.ZOOM_OUT };
}

export function filterIsEmpty(value) {
  if (value && value.values && !value.values.length) {
    return true;
  }

  if (Array.isArray(value) && !value.length) {
    return true;
  }

  if (typeof value === 'string' && !value) {
    return true;
  }

  if (typeof value === 'object') {
    const hasValue = Object.keys(value).reduce(
      (result, key) => result || Boolean(value[key]),
      false
    );
    return !hasValue;
  }

  return false;
}

export function processFilters(readOnlySearch, filters, limit, from) {
  const search = {
    filters: {},
    ...readOnlySearch,
  };
  search.filters = {};

  filters.properties.forEach(property => {
    if (!filterIsEmpty(readOnlySearch.filters[property.name]) && !property.filters) {
      search.filters[property.name] = readOnlySearch.filters[property.name];
    }

    if (property.filters) {
      const searchFilter = { ...readOnlySearch.filters[property.name] };
      property.filters.forEach(filter => {
        if (filterIsEmpty(searchFilter[filter.name])) {
          delete searchFilter[filter.name];
        }
      });

      if (Object.keys(searchFilter).length) {
        search.filters[property.name] = searchFilter;
      }
    }
  });

  search.types = filters.documentTypes;
  search.limit = limit;
  search.from = from;
  return search;
}

export function encodeSearch(_search, appendQ = true) {
  const search = { ..._search };
  Object.keys(search).forEach(key => {
    if (search[key] && search[key].length === 0) {
      delete search[key];
    }

    if (typeof search[key] === 'object' && Object.keys(search[key]).length === 0) {
      delete search[key];
    }

    if (search[key] === '') {
      delete search[key];
    }
  });

  if (search.searchTerm) {
    search.searchTerm = `${encodeURIComponent(search.searchTerm).replace(/%20/g, ' ')}:`;
  }

  const encodedSearch = rison.encode(search).replace(/searchTerm:'([^:]+):'/, "searchTerm:'$1'");
  return appendQ ? `?q=${encodedSearch}` : encodedSearch;
}

function setSearchInUrl(searchParams) {
  const { pathname } = browserHistory.getCurrentLocation();
  const path = `${pathname}/`.replace(/\/\//g, '/');
  const query = browserHistory.getCurrentLocation().query || {};

  query.q = encodeSearch(searchParams, false);

  browserHistory.push(path + toUrlParams(query));
}

export function searchDocuments(
  { search = undefined, filters = undefined },
  storeKey,
  limit = 30,
  from = 0
) {
  return (dispatch, getState) => {
    const state = getState()[storeKey];
    const currentSearch = search || state.search;
    let currentFilters = filters || state.filters;
    currentFilters = currentFilters.toJS ? currentFilters.toJS() : currentFilters;

    const searchParams = processFilters(currentSearch, currentFilters, limit, from);
    searchParams.searchTerm = state.search.searchTerm;

    const { query } = browserHistory.getCurrentLocation();
    const currentSearchParams = rison.decode(decodeURIComponent(query.q || '()'));

    if (searchParams.searchTerm && searchParams.searchTerm !== currentSearchParams.searchTerm) {
      searchParams.sort = '_score';
    }

    if (currentSearch.userSelectedSorting) {
      dispatch(actions.set(`${storeKey}.selectedSorting`, currentSearch));
    }

    searchParams.customFilters = currentSearch.customFilters;

    setSearchInUrl(searchParams);
  };
}

export function elementCreated(doc) {
  return { type: types.ELEMENT_CREATED, doc };
}

export function updateEntity(updatedDoc) {
  return { type: types.UPDATE_DOCUMENT, doc: updatedDoc };
}

export function updateEntities(updatedDocs) {
  return { type: types.UPDATE_DOCUMENTS, docs: updatedDocs };
}

export function searchSnippets(searchTerm, sharedId, storeKey) {
  return dispatch =>
    api.searchSnippets(new RequestParams({ searchTerm, id: sharedId })).then(snippets => {
      dispatch(actions.set(`${storeKey}.sidepanel.snippets`, snippets));
      return snippets;
    });
}

export function saveDocument(doc, formKey) {
  return async dispatch => {
    const updatedDoc = await documentsApi.save(new RequestParams(doc));
    dispatch(notificationActions.notify('Document updated', 'success'));
    dispatch(formActions.reset(formKey));
    dispatch(updateEntity(updatedDoc));
    dispatch(actions.updateIn('library.markers', ['rows'], updatedDoc));
    await dispatch(selectSingleDocument(updatedDoc));
  };
}

export function multipleUpdate(entities, values) {
  return async dispatch => {
    const ids = entities.map(entity => entity.get('sharedId')).toJS();
    const updatedDocs = await entitiesAPI.multipleUpdate(new RequestParams({ ids, values }));
    dispatch(notificationActions.notify('Update success', 'success'));
    dispatch(updateEntities(updatedDocs));
  };
}

export function saveEntity(entity, formModel) {
  return async dispatch => {
    const updatedDoc = await entitiesAPI.save(new RequestParams(entity));
    dispatch(formActions.reset(formModel));
    await dispatch(unselectAllDocuments());
    if (entity._id) {
      dispatch(notificationActions.notify('Entity updated', 'success'));
      dispatch(updateEntity(updatedDoc));
      dispatch(actions.updateIn('library.markers', ['rows'], updatedDoc));
    } else {
      dispatch(notificationActions.notify('Entity created', 'success'));
      dispatch(elementCreated(updatedDoc));
    }

    await dispatch(selectSingleDocument(updatedDoc));
  };
}

export function removeDocument(doc) {
  return { type: types.REMOVE_DOCUMENT, doc };
}

export function removeDocuments(docs) {
  return { type: types.REMOVE_DOCUMENTS, docs };
}

export function deleteDocument(doc) {
  return async dispatch => {
    await documentsApi.delete(new RequestParams({ sharedId: doc.sharedId }));
    dispatch(notificationActions.notify('Document deleted', 'success'));
    await dispatch(unselectAllDocuments());
    dispatch(removeDocument(doc));
  };
}

export function deleteEntity(entity) {
  return async dispatch => {
    await entitiesAPI.delete(entity);
    dispatch(notificationActions.notify('Entity deleted', 'success'));
    await dispatch(unselectDocument(entity._id));
    dispatch(removeDocument(entity));
  };
}

export function loadMoreDocuments(storeKey, amount, from) {
  return (dispatch, getState) => {
    const { search } = getState()[storeKey];
    searchDocuments({ search }, storeKey, amount, from)(dispatch, getState);
  };
}

export function getSuggestions() {
  return { type: 'GET_SUGGESTIONS' };
}

export function getDocumentReferences(sharedId, storeKey) {
  return dispatch =>
    referencesAPI.get(new RequestParams({ sharedId })).then(references => {
      dispatch(actions.set(`${storeKey}.sidepanel.references`, references));
    });
}

export function getAggregationSuggestions(storeKey, property, searchTerm) {
  const state = store.getState()[storeKey];
  const { search, filters } = state;

  const query = processFilters(search, filters.toJS(), 0);
  query.searchTerm = search.searchTerm;
  if (storeKey === 'uploads') {
    query.unpublished = true;
  }
  return searchAPI.getAggregationSuggestions(new RequestParams({ query, property, searchTerm }));
}

export function setTableViewColumns(columns) {
  return { type: types.SET_TABLE_VIEW_COLUMNS, columns };
}

export function setTableViewColumnHidden(name, hidden) {
  return {
    type: types.SET_TABLE_VIEW_COLUMN_HIDDEN,
    name,
    hidden,
  };
}

export function setTableViewAllColumnsHidden(hidden) {
  return {
    type: types.SET_TABLE_VIEW_ALL_COLUMNS_HIDDEN,
    hidden,
  };
}
