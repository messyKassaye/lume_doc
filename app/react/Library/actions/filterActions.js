import * as types from 'app/Library/actions/actionTypes';
import libraryHelper from 'app/Library/helpers/libraryFilters';
import comonPropertiesHelper from 'shared/comonProperties';
import * as libraryActions from 'app/Library/actions/libraryActions';
import prioritySortingCriteria from 'app/utils/prioritySortingCriteria';
import { actions as formActions } from 'react-redux-form';

export function filterDocumentTypes(documentTypes, storeKey) {
  return (dispatch, getState) => {
    const state = getState();

    const templates = state.templates.toJS();
    const thesauris = state.thesauris.toJS();

    let libraryFilters = comonPropertiesHelper
      .comonProperties(templates, documentTypes)
      .filter(prop => prop.filter);
    libraryFilters = libraryHelper.populateOptions(libraryFilters, thesauris);

    const usefulTemplates = documentTypes.length
      ? templates.filter(t => documentTypes.includes(t._id))
      : templates;

    const { sort, order } = prioritySortingCriteria.get({
      currentCriteria: { sort: state[storeKey].search.sort, order: state[storeKey].search.order },
      filteredTemplates: usefulTemplates.map(t => t._id),
      templates: state.templates,
      selectedSorting: state[storeKey].selectedSorting,
    });

    const search = {
      types: documentTypes,
      ...state[storeKey].search,
      sort,
      order,
    };

    const filters = { documentTypes, properties: libraryFilters };
    dispatch(libraryActions.searchDocuments({ filters, search }, storeKey));
  };
}

export function resetFilters(storeKey) {
  return (dispatch, getState) => {
    dispatch({ type: types.SET_LIBRARY_FILTERS, documentTypes: [], libraryFilters: [] });
    dispatch(
      formActions.load(`${storeKey}.search`, {
        searchTerm: '',
        filters: {},
        order: 'desc',
        sort: 'creationDate',
      })
    );
    libraryActions.searchDocuments({ search: getState()[storeKey].search }, storeKey)(
      dispatch,
      getState
    );
  };
}

export function toggleFilter(propertyName, properties) {
  return dispatch => {
    const updatedProperties = properties.map(property => {
      if (property.name === propertyName) {
        property.active = !property.active;
      }
      return property;
    });
    dispatch({ type: types.UPDATE_LIBRARY_FILTERS, libraryFilters: updatedProperties });
  };
}
