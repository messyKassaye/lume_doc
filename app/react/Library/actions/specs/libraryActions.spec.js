/**
 * @jest-environment jsdom
 */

import backend from 'fetch-mock';
import { APIURL } from 'app/config.js';
import Immutable from 'immutable';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mockID } from 'shared/uniqueID.js';
import rison from 'rison-node';
import { RequestParams } from 'app/utils/RequestParams';

import * as actions from 'app/Library/actions/libraryActions';
import * as types from 'app/Library/actions/actionTypes';
import * as notificationsTypes from 'app/Notifications/actions/actionTypes';
import { documentsApi } from 'app/Documents';
import { api } from 'app/Entities';
import { browserHistory } from 'react-router';

import referencesAPI from 'app/Viewer/referencesAPI';
import SearchApi from 'app/Search/SearchAPI';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('libraryActions', () => {
  const documentCollection = [{ name: 'Secret list of things' }];
  const aggregations = [{ prop: { buckets: [] } }];
  const templates = [{ name: 'Decision' }, { name: 'Ruling' }];
  const thesauris = [{ _id: 'abc1' }];
  let getState;

  describe('setDocuments', () => {
    it('should return a SET_DOCUMENTS action ', () => {
      const action = actions.setDocuments(documentCollection);
      expect(action).toEqual({ type: types.SET_DOCUMENTS, documents: documentCollection });
    });
  });

  describe('setTemplates', () => {
    const documentTypes = ['typea'];
    let dispatch;
    const filters = {
      documentTypes,
      properties: ['library properties'],
    };

    beforeEach(() => {
      dispatch = jasmine.createSpy('dispatch');
      getState = jasmine.createSpy('getState').and.returnValue({
        settings: {
          collection: Immutable.Map({}),
        },
        library: { filters: Immutable.fromJS(filters), search: {} },
      });
    });

    it('should dispatch a SET_LIBRARY_TEMPLATES action ', () => {
      actions.setTemplates(templates, thesauris)(dispatch, getState);
      expect(dispatch).toHaveBeenCalledWith({
        type: types.SET_LIBRARY_TEMPLATES,
        templates,
        thesauris,
      });
    });
  });

  describe('enterLibrary', () => {
    it('should return a ENTER_LIBRARY action ', () => {
      const action = actions.enterLibrary();
      expect(action).toEqual({ type: types.ENTER_LIBRARY });
    });
  });

  describe('hideFilters', () => {
    it('should return a HIDE_FILTERS action ', () => {
      const action = actions.hideFilters();
      expect(action).toEqual({ type: types.HIDE_FILTERS });
    });
  });

  describe('showFilters', () => {
    it('should return a SHOW_FILTERS action ', () => {
      const action = actions.showFilters();
      expect(action).toEqual({ type: types.SHOW_FILTERS });
    });
  });

  describe('setPreviewDoc', () => {
    it('should return a SET_PREVIEW_DOC action ', () => {
      const action = actions.setPreviewDoc('id');
      expect(action).toEqual({ type: types.SET_PREVIEW_DOC, docId: 'id' });
    });
  });

  describe('overSuggestions', () => {
    it('should return a OVER_SUGGESTIONS action ', () => {
      const action = actions.setOverSuggestions(true);
      expect(action).toEqual({ type: types.OVER_SUGGESTIONS, hover: true });
    });
  });

  describe('encodeSearch', () => {
    it('should return a query string with ?q= at the beginning by default', () => {
      expect(actions.encodeSearch({ a: 1, b: 'z' })).toBe('?q=(a:1,b:z)');
    });

    it('should encode special characters to avoid errors with rison', () => {
      expect(actions.encodeSearch({ searchTerm: 'number 19# (/) "', b: 'z' })).toBe(
        "?q=(b:z,searchTerm:'number 19%23 (%2F) %22')"
      );
    });
    it('should allow returning a rison query value, not appending the ?q= when other options may be found in the URL', () => {
      expect(actions.encodeSearch({ a: 1, b: 'z' }, false)).toBe('(a:1,b:z)');
    });

    it('should keep the search term inside single quotes to avoid errors with rison', () => {
      const toCode = {
        searchTerm: 'a:b',
        sort: '_score',
        order: 'desc',
        includeUnpublished: false,
        unpublished: false,
        allAggregations: false,
      };
      const encodedSearch = actions.encodeSearch(toCode, false);
      const coded =
        "(allAggregations:!f,includeUnpublished:!f,order:desc,searchTerm:'a%3Ab',sort:_score,unpublished:!f)";
      expect(encodedSearch).toBe(coded);
    });
  });

  describe('Zoom functions', () => {
    it('should zoom in and out', () => {
      expect(actions.zoomIn()).toEqual({ type: types.ZOOM_IN });
      expect(actions.zoomOut()).toEqual({ type: types.ZOOM_OUT });
    });
  });

  describe('async action', () => {
    let dispatch;
    beforeEach(() => {
      backend.restore();
      backend
        .get(`${APIURL}search?searchTerm=batman`, { body: JSON.stringify(documentCollection) })
        .get(
          `${APIURL}search?searchTerm=batman` +
            '&filters=%7B%22author%22%3A%7B%22value%22%3A%22batman%22%2C%22type%22%3A%22text%22%7D%7D' +
            '&aggregations=%5B%5D' +
            '&types=%5B%22decision%22%5D',
          { body: JSON.stringify({ rows: documentCollection, aggregations }) }
        )
        .get(
          `${APIURL}search?searchTerm=batman&filters=%7B%7D&aggregations=%5B%5D&types=%5B%22decision%22%5D`,
          { body: JSON.stringify({ rows: documentCollection, aggregations }) }
        );
      dispatch = jasmine.createSpy('dispatch');
    });

    afterEach(() => backend.restore());

    describe('searchDocuments', () => {
      let store;
      let state;
      const storeKey = 'library';
      beforeEach(() => {
        state = {
          properties: [
            { name: 'author' },
            { name: 'inactive' },
            { name: 'date', type: 'date' },
            { name: 'select', type: 'select' },
            { name: 'multiselect', type: 'multiselect' },
            {
              name: 'nested',
              type: 'nested',
              nestedProperties: [{ key: 'prop1', label: 'prop one' }],
            },
            {
              name: 'relationshipfilter',
              type: 'relationshipfilter',
              filters: [
                { name: 'status', type: 'select' },
                { name: 'empty', type: 'date' },
              ],
            },
          ],
          documentTypes: ['decision'],
        };
        store = {
          settings: {
            collection: Immutable.Map({}),
          },
          library: {
            filters: Immutable.fromJS(state),
            search: {
              searchTerm: 'batman',
              customFilters: { property: { values: ['value'] } },
              filters: {},
            },
          },
        };
        spyOn(browserHistory, 'getCurrentLocation').and.returnValue({
          pathname: '/library',
          query: { view: 'chart' },
          search: '?q=()',
        });
        getState = jasmine.createSpy('getState').and.returnValue(store);
      });

      it('should convert the search and set it to the url query based on filters on the state', () => {
        const search = {
          searchTerm: 'batman',
          filters: {
            author: 'batman',
            date: 'dateValue',
            select: 'selectValue',
            multiselect: 'multiValue',
            nested: 'nestedValue',
            relationshipfilter: { status: { values: ['open'] }, empty: '' },
          },
        };

        const expectedQuery = {
          filters: {
            author: 'batman',
            date: 'dateValue',
            multiselect: 'multiValue',
            nested: 'nestedValue',
            relationshipfilter: { status: { values: ['open'] } },
            select: 'selectValue',
          },
          limit: 30,
          from: 0,
          searchTerm: 'batman',
          sort: '_score',
          types: ['decision'],
        };

        spyOn(browserHistory, 'push');
        actions.searchDocuments({ search }, storeKey, 30)(dispatch, getState);
        let queryObject = rison.decode(
          browserHistory.push.calls.mostRecent().args[0].split('q=')[1]
        );
        expect(queryObject).toEqual(expectedQuery);

        search.filters.relationshipfilter.status.values = [];
        actions.searchDocuments({ search }, storeKey, 'limit')(dispatch, getState);
        queryObject = rison.decode(browserHistory.push.calls.mostRecent().args[0].split('q=')[1]);
        expect(queryObject.filters.relationshipfilter).not.toBeDefined();
      });

      it('should use passed filters when passed', () => {
        const search = {
          searchTerm: 'batman',
          filters: {
            author: 'batman',
            date: { from: null },
            select: 'selectValue',
            multiselect: { values: [] },
            nested: 'nestedValue',
            object: {},
          },
        };

        const { filters } = store.library;

        const limit = 60;
        spyOn(browserHistory, 'push');
        actions.searchDocuments({ search, filters }, storeKey, limit)(dispatch, getState);

        expect(browserHistory.push).toHaveBeenCalledWith(
          "/library/?view=chart&q=(filters:(author:batman,nested:nestedValue,select:selectValue),from:0,limit:60,searchTerm:'batman',sort:_score,types:!(decision))" //eslint-disable-line
        );
      });

      it('should use customFilters from the current search on the store', () => {
        const limit = 60;
        spyOn(browserHistory, 'push');
        actions.searchDocuments({}, storeKey, limit)(dispatch, getState);

        expect(browserHistory.push).toHaveBeenCalledWith(
          "/library/?view=chart&q=(customFilters:(property:(values:!(value))),filters:(),from:0,limit:60,searchTerm:'batman',sort:_score,types:!(decision))" //eslint-disable-line
        );
      });

      it('should set the storeKey selectedSorting if user has selected a custom sorting', () => {
        const expectedDispatch = {
          type: 'library.selectedSorting/SET',
          value: { searchTerm: 'batman', filters: { author: 'batman' }, userSelectedSorting: true },
        };
        spyOn(browserHistory, 'push');
        actions.searchDocuments(
          {
            search: {
              searchTerm: 'batman',
              filters: { author: 'batman' },
              userSelectedSorting: true,
            },
          },
          storeKey
        )(dispatch, getState);
        expect(dispatch).toHaveBeenCalledWith(expectedDispatch);
      });

      it('should set sort by relevance when the search term has changed and has value', () => {
        browserHistory.getCurrentLocation.and.returnValue({
          pathname: '/library',
          query: { view: 'chart' },
          search: '?q=(searchTerm:%27batman%20begings%27)',
        });
        spyOn(browserHistory, 'push');
        actions.searchDocuments(
          { search: { searchTerm: 'batman' }, filters: { properties: [] } },
          storeKey
        )(dispatch, getState);
        expect(browserHistory.push).toHaveBeenCalledWith(
          "/library/?view=chart&q=(from:0,limit:30,searchTerm:'batman',sort:_score)"
        );
      });

      it('should respect the sorting criteria when the search term has not changed and has value', () => {
        browserHistory.getCurrentLocation.and.returnValue({
          pathname: '/library',
          query: { q: '(searchTerm:batman,sort:_score)' },
          search: '?q=(searchTerm:%27batman%20begings%27)',
        });
        spyOn(browserHistory, 'push');
        actions.searchDocuments(
          {
            search: { searchTerm: 'batman', sort: 'title', order: 'desc' },
            filters: { properties: [] },
          },
          storeKey
        )(dispatch, getState);
        expect(browserHistory.push).toHaveBeenCalledWith(
          "/library/?q=(from:0,limit:30,order:desc,searchTerm:'batman',sort:title)"
        );
      });
    });

    describe('saveDocument', () => {
      it('should save the document and dispatch a notification on success', done => {
        mockID();
        spyOn(documentsApi, 'save').and.returnValue(Promise.resolve('response'));
        spyOn(referencesAPI, 'get').and.returnValue(Promise.resolve('response'));
        const doc = { name: 'doc' };

        const expectedActions = [
          {
            type: notificationsTypes.NOTIFY,
            notification: { message: 'Document updated', type: 'success', id: 'unique_id' },
          },
          { type: 'rrf/reset', model: 'library.sidepanel.metadata' },
          { type: types.UPDATE_DOCUMENT, doc: 'response' },
          { type: 'library.markers/UPDATE_IN', key: ['rows'], value: 'response' },
          { type: types.SELECT_SINGLE_DOCUMENT, doc: 'response' },
        ];
        const store = mockStore({});

        store
          .dispatch(actions.saveDocument(doc, 'library.sidepanel.metadata'))
          .then(() => {
            expect(documentsApi.save).toHaveBeenCalledWith(new RequestParams(doc));
            expect(store.getActions()).toEqual(expectedActions);
          })
          .then(done)
          .catch(done.fail);
      });
    });

    describe('multipleUpdate', () => {
      it('should update selected entities with the given metadata', async () => {
        mockID();
        const metadataResponse = { text: 'something new', rest_of_metadata: 'test' };
        spyOn(api, 'multipleUpdate').and.returnValue(
          Promise.resolve([
            { sharedId: '1', metadataResponse },
            { sharedId: '2', metadataResponse },
          ])
        );
        const entities = Immutable.fromJS([{ sharedId: '1' }, { sharedId: '2' }]);

        const expectedActions = [
          {
            type: notificationsTypes.NOTIFY,
            notification: { message: 'Update success', type: 'success', id: 'unique_id' },
          },
          {
            type: types.UPDATE_DOCUMENTS,
            docs: [
              { sharedId: '1', metadataResponse },
              { sharedId: '2', metadataResponse },
            ],
          },
        ];
        const store = mockStore({});
        const changes = { property_changes: 'change' };
        await store.dispatch(actions.multipleUpdate(entities, { metadata: changes }));
        expect(api.multipleUpdate).toHaveBeenCalledWith(
          new RequestParams({ ids: ['1', '2'], values: { metadata: changes } })
        );
        expect(store.getActions()).toEqual(expectedActions);
      });
    });

    describe('deleteDocument', () => {
      it('should delete the document and dispatch a notification on success', done => {
        mockID();
        spyOn(documentsApi, 'delete').and.returnValue(Promise.resolve('response'));
        const doc = { sharedId: 'sharedId', name: 'doc' };

        const expectedActions = [
          {
            type: notificationsTypes.NOTIFY,
            notification: { message: 'Document deleted', type: 'success', id: 'unique_id' },
          },
          { type: types.UNSELECT_ALL_DOCUMENTS },
          { type: types.REMOVE_DOCUMENT, doc },
        ];
        const store = mockStore({});

        store
          .dispatch(actions.deleteDocument(doc))
          .then(() => {
            expect(documentsApi.delete).toHaveBeenCalledWith(
              new RequestParams({ sharedId: doc.sharedId })
            );
            expect(store.getActions()).toEqual(expectedActions);
          })
          .then(done)
          .catch(done.fail);
      });
    });

    describe('searchSnippets', () => {
      it('should search snippets for the searchTerm', done => {
        spyOn(SearchApi, 'searchSnippets').and.returnValue(Promise.resolve('response'));

        const expectedActions = [{ type: 'storeKey.sidepanel.snippets/SET', value: 'response' }];

        const store = mockStore({ locale: 'es' });

        store
          .dispatch(actions.searchSnippets('query', 'sharedId', 'storeKey'))
          .then(snippets => {
            expect(snippets).toBe('response');
            expect(SearchApi.searchSnippets).toHaveBeenCalledWith(
              new RequestParams({ searchTerm: 'query', id: 'sharedId' })
            );
            expect(store.getActions()).toEqual(expectedActions);
          })
          .then(done)
          .catch(done.fail);
      });
    });

    describe('getDocumentReferences', () => {
      it('should set the library sidepanel references', done => {
        mockID();
        spyOn(referencesAPI, 'get').and.returnValue(Promise.resolve('referencesResponse'));

        const expectedActions = [
          { type: 'library.sidepanel.references/SET', value: 'referencesResponse' },
        ];

        const store = mockStore({ locale: 'es' });

        store
          .dispatch(actions.getDocumentReferences('id', 'library'))
          .then(() => {
            expect(referencesAPI.get).toHaveBeenCalledWith(new RequestParams({ sharedId: 'id' }));
            expect(store.getActions()).toEqual(expectedActions);
          })
          .then(done)
          .catch(done.fail);
      });
    });

    describe('selectDocument', () => {
      describe('when the doc has not semantic search but the active sidepanel tab is semantic search', () => {
        it('should reset the active sidepanel tab', async () => {
          const doc = { sharedId: 'doc' };

          const store = mockStore({ library: { sidepanel: { tab: 'semantic-search-results' } } });
          await store.dispatch(actions.selectDocument(doc));
          expect(store.getActions()).toMatchSnapshot();
        });
      });
    });
  });

  describe('saveEntity', () => {
    it('should save the document and dispatch a notification on success', done => {
      mockID();
      spyOn(api, 'save').and.returnValue(Promise.resolve('response'));
      spyOn(referencesAPI, 'get').and.returnValue(Promise.resolve('response'));
      const doc = { name: 'doc' };

      const expectedActions = [
        { model: 'library.sidepanel.metadata', type: 'rrf/reset' },
        { type: 'UNSELECT_ALL_DOCUMENTS' },
        {
          notification: {
            id: 'unique_id',
            message: 'Entity created',
            type: 'success',
          },
          type: 'NOTIFY',
        },
        { doc: 'response', type: 'ELEMENT_CREATED' },
        { doc: 'response', type: 'SELECT_SINGLE_DOCUMENT' },
      ];
      const store = mockStore({});

      store
        .dispatch(actions.saveEntity(doc, 'library.sidepanel.metadata'))
        .then(() => {
          expect(api.save).toHaveBeenCalledWith(new RequestParams(doc));
          expect(store.getActions()).toEqual(expectedActions);
        })
        .then(done)
        .catch(done.fail);
    });
  });
});
