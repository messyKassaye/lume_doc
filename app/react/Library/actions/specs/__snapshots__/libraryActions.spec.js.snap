// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`libraryActions async action selectDocument when the doc has not semantic search but the active sidepanel tab is semantic search should reset the active sidepanel tab 1`] = `
Array [
  Object {
    "type": "library.sidepanel.tab/SET",
    "value": "",
  },
  Object {
    "doc": Object {
      "sharedId": "doc",
    },
    "type": "SELECT_DOCUMENT",
  },
]
`;
